File: solutions/twisty-little-passages/kotlin/twistylittlepassages.kt
fun read(): Pair<Int, Int> { // Reading inputs
    val (r, p) = readLine()!!.split(" ").map { it.toInt() }
    return r to p
}

fun move(Move: String) {     //Printing action to do (walk, teleport or estimate)
    println(Move)
    System.out.flush()       //Flushing output stream
}

fun main() {
    var cases = readLine()!!.toInt()
    repeat(cases) {          // Iterate tru all cases
        val (N, K) = read()  // N: number of rooms, K: operations to solve case
        val (R, P) = read()  // R: current room and P: passages in room
        
        // Empty set
        var roomsLeft = hashSetOf<Int>() 
        // sSt {1, 2, ..., N}
        for (i in 1 .. N) roomsLeft.add(i)
        // Removes room R from the set
        if (R in roomsLeft) roomsLeft.remove(R)
        
        // Rooms visited degree sum
        var degree = 0L
        
        // Degree and count of every teleported room
        var degreeRL = 0L
        var countT = 1
        
        degree = P.toLong()
        degreeRL = degree

        repeat(K / 2) { // Iterate tru all the operations given for the case
            // Odd cases
            move("T ${roomsLeft.iterator().next()}")
            val (RT, PT) = read()
            degreeRL += PT
            countT++
            if (RT in roomsLeft) {
                roomsLeft.remove(RT)
                degree += PT
            }
            // Even cases
            move("W")
            val (RW, PW) = read()
            if (RW in roomsLeft) {
                roomsLeft.remove(RW)
                degree += PW
            }
        }
        // Degree Avg of every visited room using tp
        val degreeAvg = degreeRL / countT
        
        // Graph formula. Sum of all degrees / 2
        // Sum of degree of rooms visited + Avg degree times number of rooms left
        val result = (((degree + (degreeAvg * roomsLeft.size)) / 2)).toLong()
        move("E $result")
    }
}

File: solutions/twisty-little-passages/python/twistylittlepassages.py
def main():
    cases = int(input())

    for case in range(cases):
        solve()

def solve():
    # N: number of rooms, K: number of tries to solve case
    N, K = list(map(int, input().split()))
    # R: current room, P: number of passages seen from room
    R, P = list(map(int, input().split()))
    
    # set {1, 2, ..., N + 1}
    rooms_left = set(range(1, N + 1))

    # removes room R from the set
    if R in rooms_left:
        rooms_left.remove(R)

    # degree of room
    degree = P

    # degree and count of every teleported room
    degree_vt = degree
    count_vt = 1


    for i in range(K):
        # if there's no room left break
        if not rooms_left:
            break

        # in even cases
        if i % 2 == 0:
            print(f'W')
            R, P = list(map(int, input().split()))

        # in odd cases
        else:
            print(f'T {next(iter(rooms_left))}')
            R, P = list(map(int, input().split()))

            degree_vt += P
            count_vt += 1

        if R in rooms_left:
            rooms_left.remove(R)
            degree += P
    
    # degrees average of every visited room
    degree_avg = degree_vt / count_vt

    # sum of degree of rooms visited and average degree times
    # number of rooms left
    result = int((degree + degree_avg * len(rooms_left)) / 2)
    print(f'E {result}')

if __name__ == "__main__":
    main()
